Product Filters: gracefully handle invalid items

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

We don't handle invalid term well which throws a warning when one is used. This PR fix that issue by only process valid term.  I also add `hide_empty` argument back to the query clause which was removed in https://github.com/woocommerce/woocommerce/pull/59508 to ensure the final query doesn't change.

Fixes [WOOPLUG-4978](https://linear.app/a8c/issue/WOOPLUG-4978/handle-invalid-product-filter-terms-gracefully)

<!-- Begin testing instructions -->

| Before | After |
|--------|--------|
| <img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/6c727aef-340f-4ecc-85fb-25a6f7870f23" /> | <img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/cb525814-d2e3-4aa7-90ed-47ae37108d32" /> |
| <img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/63389899-c0c9-4bc0-bcf5-ccb22b1fd34e" /> | <img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/6db4a000-3627-48eb-86ab-db73b26cbbd7" /> |
| <img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/eef4d4fe-7040-4698-8d2a-b5493457d311" /> | <img width="1025" height="769" alt="image" src="https://github.com/user-attachments/assets/e77f1bd8-ed89-4993-8f72-12b7f79da353" /> |

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Given that the current store has a Color attribute.
2. Go to shop page on the front end, see manually add `filter_color=invalid-color` to the URL.
3. See no notice/warning and Active Filters doesn't contain invalid item.
4. Try invalid rating_filter value like the screenshot, see Active Filters doesn't contain invalid item.

Note that the result with invalid items of Price Filter and Rating Filter doesn't look right, as it should be empty result with invalid filter. But we're relying on legacy behavior that has been introduced since the age of classic themes and widgets. We will address them in the future.

<!-- End testing instructions -->

dinhtungdu committed:
fix: handle the case when some selected terms are empty

dinhtungdu committed:
refactor: better preparing selected attributes

dinhtungdu committed:
refactor: more validation and early return

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
fix: validate rating values

dinhtungdu committed:
fix: remove init methods of QueryFilters

dinhtungdu committed:
fix: handle invalid stock items

dinhtungdu committed:
style: fix linting issues in ProductFilterAttribute and QueryFilters classes

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
```diff
@@ -70,9 +70,15 @@ public function prepare_selected_filters( $items, $params ) {
+				return $rating >= 1 && $rating <= 5;
+			}
 		);
+		$active_ratings = array_unique( $active_ratings );
```
I think it's minor and super edge case but mapping values to `intval` allows you to keep fractions in url which are then treated as rating... but their value is different so you can add the same rating one more time. Like I said, it's an edge case but may lead to some unexpected situations:

URL: `shop/?rating_filter=4.1234,5.1234`
What happens: products are filtered correctly and active filter is updated. Rating filter appears as no filter is applied
<img width="1500" height="704" alt="image" src="https://github.com/user-attachments/assets/1661ee7f-b82d-4c72-b61e-c9e1882f6be9" />

I'm clicking rating 4 and 5:

URL is fixed: `/?rating_filter=5%2C4`
What happens: products are filtered correctly, active filter shows duplicated result

<img width="1530" height="768" alt="image" src="https://github.com/user-attachments/assets/3c6b7ad4-670a-413f-a19b-c83abdb13091" />

dinhtungdu replied:
Good catch! This is something we need to fix IMO. I pushed 3765c34 to fix this. Let me know if it looks good to you.

kmanijak replied:
So while the active filter doesn't display the value, products are actually filtered 😄 check the second and third image, product is filtered out but no active filters is present.

URL: `shop` | URL: `/shop/?rating_filter=5.1234` | URL: `/shop/?rating_filter=5
-- | -- | --
<img width="1594" height="1196" alt="image" src="https://github.com/user-attachments/assets/15025279-367c-4f7e-96ce-8dbc29bc1dd6" /> | <img width="1575" height="621" alt="image" src="https://github.com/user-attachments/assets/f6b49fb1-63ac-4b89-9aa8-5bf75354eaaf" /> |  <img width="1589" height="715" alt="image" src="https://github.com/user-attachments/assets/2da1d2be-6c1d-4746-b0c8-f5b4e129f551" />

Of course it's an edge case and I don't think it's significantly impacts the UX (why would anyone put fractions 😅)

dinhtungdu replied:
hmm, I thought I fixed this already, let me see.

dinhtungdu replied:
Here is what I got on my end. Can you check it again? 
<img width="1537" height="850" alt="image" src="https://github.com/user-attachments/assets/9c5333e9-441c-4133-8891-ea4d945c788e" />

kmanijak replied:
But if I get it right, you have no products with rating 5? Try with `4.123`

dinhtungdu replied:
Ah right, I see.  The rating filter for main query is actually handled in `class-wc-query.php`
```
			$rating_filter = array_filter( array_map( 'absint', explode( ',', wp_unslash( $_GET['rating_filter'] ) ) ) );
```
I can refactor it to make it match that behavior. Let me think. It's the problem of matching current products and count.

dinhtungdu replied:
@kmanijak I think this is good now: 
<img width="1537" height="850" alt="image" src="https://github.com/user-attachments/assets/22615ef4-cb1d-4ade-b3d7-b4684f3aab72" />

kmanijak replied:
Now there's something wrong with rating filter. One filter overrides the other in the URL and hence it leads to inconsistent results:

https://d.pr/i/emF24s

Basically when clicking more than one rating, for example 3 and 4:
- expected URL `shop/?rating_filter=3` -> `shop/?rating_filter=3%2C4`
- actual URL `shop/?rating_filter=3` -> `shop/?rating_filter=4`

dinhtungdu replied:
It's not an edge case. The filter option value must be a string for URL manipulation to work as expected. I forgot this and changed the type of the rating value to a number. I fixed it, and the filter now works as before, including fraction handling.

kmanijak reviewed (commented):
Nice improvements! I found one "edge case" with rating filter described in inline comment. But I just realised it's not exactly related to fractions in URL.

If I have URL `shop/?rating_filter=5%2C4` and refresh the page, filter is applied but rating filter is not updated (no checkboxes) and I can click the filter again, similar as in the inline comment:

<img width="1494" height="783" alt="image" src="https://github.com/user-attachments/assets/2cd7d29c-52aa-4f87-ba87-83d8a9008953" />

Let mw know if you want to fix the issue in the same PR or consider it a separate one (and I'm happy to approve then).

dinhtungdu committed:
fix: force to use valid rating value

dinhtungdu committed:
style: fix equals sign alignment in ProductFilterRating

Fix alignment of equals sign in $valid_ratings assignment to match
surrounding assignments in ProductFilterRating.php:73

kmanijak reviewed (approved):
It's looking good! I left one comment but I think it's so edge case, it may not be even worth spending time on 😅 

I tested couple of different incorrect filters and all seems to be working fine!

dinhtungdu committed:
fix: handle fractions

dinhtungdu committed:
fix: option value must be string

dinhtungdu commented:
> It's looking good! I left one comment but I think it's so edge case, it may not be even worth spending time on 😅

@kmanijak I'm sorry, the type of the option value is the cause of the issue here. I changed it from string to number, but it needs to be a string. That is why we have all the problems we discussed. It's not an edge case. I fixed the type, and everything should work as expected now. Can you please take a (hopefully) final look at this?

kmanijak reviewed (approved):
> It's not an edge case.

I meant putting the fractions in the URL not that one filter overrides another 😄 but you're right!

I re-tested the PR and seems like all's good now!
- invalid items are handled gracefully
- I didn't spot any weird quirks anymore!

Good job, appreciate the responsiveness!
